File: rack/drivers/ChassisProxy.java
/*
 * RACK - Robotics Application Construction Kit
 * Copyright (C) 2005-2006 University of Hannover
 *                         Institute for Systems Engineering - RTS
 *                         Professor Bernardo Wagner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * Authors
 *      Joerg Langenberg <joerg.langenberg@gmx.net>
 *
 */
package rack.drivers;

import rack.main.*;
import rack.main.tims.*;


public class ChassisProxy extends RackDataProxy
{
    public static final byte MSG_CHASSIS_MOVE =
        RackProxy.MSG_POS_OFFSET + 1;

    public static final byte MSG_CHASSIS_GET_PARAMETER =
        RackProxy.MSG_POS_OFFSET + 2;

    public static final byte MSG_CHASSIS_SET_ACTIVE_PILOT =
        RackProxy.MSG_POS_OFFSET + 3;

    public static final byte MSG_CHASSIS_PARAMETER =
        RackProxy.MSG_NEG_OFFSET - 1;

    public static final byte INVAL_PILOT = -1;

    public ChassisProxy(int id, TimsMbx replyMbx)
    {
        super(RackName.create(RackName.CHASSIS, id), replyMbx, 5000, 1000, 1000);
        this.id = id;
    }

    public synchronized ChassisDataMsg getData(int recordingtime)
    {
        try
        {
            TimsDataMsg raw = getRawData(recordingtime);
            if (raw != null)
            {
                ChassisDataMsg data = new ChassisDataMsg(raw);
                return data;
            }
            else
            {
                return null;
            }
        }
        catch (TimsException e)
        {
            System.out.println(e.toString());
            return null;
        }
    }

    public synchronized ChassisDataMsg getData()
    {
        return (getData(0));
    }

    public synchronized ChassisParamMsg getParam()
    {
        currentSequenceNo++;
        try
        {
            replyMbx.send0(MSG_CHASSIS_GET_PARAMETER, commandMbx,
                    (byte) 0, currentSequenceNo);

            TimsDataMsg reply;
            do
            {
                reply = replyMbx.receive(1000);
            }
            while ((reply.seqNr != currentSequenceNo)
                    & (reply.type == MSG_CHASSIS_PARAMETER));
            
            ChassisParamMsg data = new ChassisParamMsg(reply);
            return (data);
        }
        catch (TimsException e)
        {
        	System.out.println(e);
            return (null);
        }
    }

    public synchronized void setActivePilot(int pilotMbx)
    {
        currentSequenceNo++;
        try
        {
            ChassisSetActivePilotMsg cmdMsg = new ChassisSetActivePilotMsg(
                    pilotMbx);

            replyMbx.send(MSG_CHASSIS_SET_ACTIVE_PILOT, commandMbx,
                          (byte) 0, currentSequenceNo, cmdMsg);

            TimsDataMsg reply;
            do
            {
                reply = replyMbx.receive(onTimeout);
            }
            while (reply.seqNr != currentSequenceNo);

            if (reply.type == RackProxy.MSG_OK)
            {
                System.out.println(RackName.nameString(replyMbx.getName()) + ": "
                        + RackName.nameString(commandMbx) + ".setActivePilot");
            }
            else
            {
                System.out.println(RackName.nameString(replyMbx.getName()) + ": "
                        + RackName.nameString(commandMbx)
                        + ".setActivePilot replied error");
            }
        }
        catch (TimsException e)
        {
            System.out.println(RackName.nameString(replyMbx.getName()) + ": "
                    + RackName.nameString(commandMbx) + ".on " + e);
        }
    }

    public int getCommandMbx()
    {
        return (RackName.create(RackName.CHASSIS, id));
    }
}

File: rack/drivers/chassis_proxy.cpp
/*
 * RACK - Robotics Application Construction Kit
 * Copyright (C) 2005-2006 University of Hannover
 *                         Institute for Systems Engineering - RTS
 *                         Professor Bernardo Wagner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * Authors
 *      Oliver Wulf <wulf@rts.uni-hannover.de>
 *
 */
#include <drivers/chassis_proxy.h>

//
// proxy functions
//

int ChassisProxy::getData(chassis_data *recv_data, ssize_t recv_datalen,
                          rack_time_t timeStamp, uint64_t reply_timeout_ns)
{
    message_info msgInfo;

    int ret = RackDataProxy::getData((void *)recv_data, recv_datalen, timeStamp,
                                    reply_timeout_ns, &msgInfo);
    if (ret)
    {
        return ret;
    }

    recv_data = ChassisData::parse(&msgInfo);
    return 0;
}

int ChassisProxy::move(int32_t vx, int32_t vy, float omega, uint64_t reply_timeout_ns)
{
    chassis_move_data send_data;
    send_data.vx    = vx;
    send_data.vy    = vy;
    send_data.omega = omega;

    return proxySendDataCmd(MSG_CHASSIS_MOVE, &send_data,
                            sizeof(chassis_move_data), reply_timeout_ns);
}

int ChassisProxy::getParam(chassis_param_data *recv_data, ssize_t recv_datalen,
                           uint64_t reply_timeout_ns)
{
    message_info msgInfo;

    int ret = proxyRecvDataCmd(MSG_CHASSIS_GET_PARAMETER, MSG_CHASSIS_PARAMETER,
                              (void *)recv_data, recv_datalen,
                              reply_timeout_ns, &msgInfo);
    if (ret)
    {
        return ret;
    }

    recv_data = ChassisParamData::parse(&msgInfo);
    return 0;
}

int ChassisProxy::setActivePilot(uint32_t pilotMbxAdr, uint64_t reply_timeout_ns)
{
    chassis_set_active_pilot_data send_data;
    send_data.activePilot = pilotMbxAdr;

    return proxySendDataCmd(MSG_CHASSIS_SET_ACTIVE_PILOT, &send_data,
                            sizeof(chassis_set_active_pilot_data),
                            reply_timeout_ns);
}
